### PR TITLE
Optimisation of data processing for actionX and output

### PR DESCRIPTION
**Pull request by hnil**
* removed double evaluation of `processLocalData`
* separated fluid in place work
* removed double evaluation of `processLocalData` from other cell based data

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

File: ebos/eclwriter.hh [hunk 1]
```diff
@@ -507,6 +515,7 @@ private:
     void prepareLocalCellData(const bool isSubStep,
                               const int  reportStepNum)
     {
+        if( !(eclOutputModule_->localDataValid()) ){
```
Comment (bska) [inline ebos/eclwriter.hh:518]: You might alternatively write this as
```C++
if (eclOutputModule_->localDataValid()) {
    return;
}
```
which would save one level of nesting.

File: ebos/eclwriter.hh [hunk 2]
```diff
@@ -470,7 +478,7 @@ public:
     void endRestart()
     {}
 
-    const EclOutputBlackOilModule<TypeTag>& eclOutputModule() const
+    EclOutputBlackOilModule<TypeTag>& eclOutputModule() const
```
Comment (bska) [inline ebos/eclwriter.hh:481]: I think I'd prefer if this were left to return a reference to `const`, and then we'd add a new function, e.g.,
```C++
EclOutputBlackoilModule<TypeTag>& mutableEclOutputModule()
{ return *eclOutputModule_; }
```
instead.  At least in that way, it's clear at the call site that we're performing a potentially mutating operation on the output module object.  You'd be forced to write something like
```C++
ebosSimulator_.problem().eclWriter()->mutableEclOutputModule().invalidateLocalData();
```
instead.

**Review by bska — COMMENTED** [review]
Apart from a couple of minor issues I like this a lot and I really look forward to seeing the benchmark results later.  I think we'll nevertheless merge this more or less as is.

File: ebos/eclwriter.hh (hunk 1 above)
Comment (hnil) [reply ebos/eclwriter.hh:518]: changed it.

File: ebos/eclwriter.hh (hunk 2 above)
Comment (hnil) [reply ebos/eclwriter.hh:481]: changed it.

**Comment by hnil** [comment]
Test on norne. original code 60 s on post and pre + 20 on output. With pullrequest 40 + 1.
På SPE9_CP original code 1.56 + 0.7 til 1.5 + 0.14.

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now.  I'll just run one more build check and then I'll merge this once the check is clear.

**Comment by bska** [comment]
> I'll just run one more build check and then I'll merge this once the check is clear.

Build is successful.  I'll merge this into the master branch now.  Thanks a lot for improving the performance here!

**Pull request merged**